### PR TITLE
Decouple XContentBuilder from BytesReference

### PR DESCRIPTION
**Pull request by dakrone**
This commit removes all mentions of `BytesReference` from `XContentBuilder`.
This is needed so that we can completely decouple the XContent code and move it
into its own dependency.

While this change appears large, it is due to two main changes, moving
`.bytes()` and `.string()` out of XContentBuilder itself into static methods
`BytesReference.bytes` and `Strings.toString` respectively. The rest of the
change is code reacting to these changes (the majority of it in tests).

Relates to #28504

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 1]
```diff
@@ -81,7 +81,6 @@ private void doFromXContentTestWithRandomFields(boolean addRandomFields) throws
             parsedGetResponse = GetResponse.fromXContent(parser);
             assertNull(parser.nextToken());
         }
-        assertEquals(expectedGetResponse, parsedGetResponse);
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: Can you explain why we need this line?

Comment (nik9000) [reply server/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: Or, rather, need to remove it?

File: server/src/test/java/org/elasticsearch/common/xcontent/builder/XContentBuilderTests.java [hunk 2]
```diff
@@ -93,61 +95,61 @@ public void testRaw() throws IOException {
...
         {
             XContentBuilder xContentBuilder = XContentFactory.contentBuilder(XContentType.JSON);
             xContentBuilder.startObject();
             xContentBuilder.field("test", "value");
-            xContentBuilder.rawField("foo", new BytesArray("{\"test\":\"value\"}"));
+            xContentBuilder.rawField("foo", new BytesArray("{\"test\":\"value\"}").streamInput());
             xContentBuilder.field("test1", "value1");
             xContentBuilder.endObject();
-            assertThat(xContentBuilder.bytes().utf8ToString(), equalTo("{\"test\":\"value\",\"foo\":{\"test\":\"value\"},\"test1\":\"value1\"}"));
+            assertThat(BytesReference.bytes(xContentBuilder).utf8ToString(), equalTo("{\"test\":\"value\",\"foo\":{\"test\":\"value\"},\"test1\":\"value1\"}"));
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/common/xcontent/builder/XContentBuilderTests.java]: These should probably move into xcontent's tests rather than server's tests so it won't be able to use this.

But that can be another PR :).

It does make me wonder if `string` should be implemented in the xcontent artifact without Lucene.

Comment (nik9000) [reply server/src/test/java/org/elasticsearch/common/xcontent/builder/XContentBuilderTests.java]: I think these should use `Strings.toString(xContentBuilder)` at least.

File: server/src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractXContentFilteringTestCase.java [hunk 3]
```diff
@@ -69,18 +70,18 @@ public void testSingleFieldObject() throws IOException {
     }
 
     static void assertXContentBuilderAsString(final XContentBuilder expected, final XContentBuilder actual) {
-        assertThat(actual.bytes().utf8ToString(), is(expected.bytes().utf8ToString()));
+        assertThat(BytesReference.bytes(actual).utf8ToString(), is(BytesReference.bytes(expected).utf8ToString()));
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractXContentFilteringTestCase.java]: I think these should use `Strings.toString(actual)`

File: server/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 4]
```diff
@@ -173,6 +170,13 @@ public XContentType contentType() {
         return generator.contentType();
     }
 
+    /**
+     * @return the output stream to which the built object is being written
```
[the remark this answers is not in the document]
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: There isn't a way that I see to assert that the OutputStream is already closed (other than closing it). I'll add documentation about the modification though.

File: server/src/test/java/org/elasticsearch/action/get/GetResponseTests.java (hunk 1 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: That was a mistake, I will re-add this.

**Comment by dakrone** [comment]
Thanks for taking a look @nik9000, I pushed commits addressing your comments

**Review by nik9000 — APPROVED** [review]
